keps/sig-storage/2639-secret-configmap-protection/README.md

```diff
@@ -0,0 +1,652 @@
+    2. Check API server if the `Secret` is actually not used by any of the resources, and if not used, delete `kubernetes.io/secret-protection` lien from the `Secret`
+
+- ConfigMap protection controller:
+  - The controller watches `Pod`, `Deployment`, `StatefulSet`, `DaemonSet`, and `CronJob`
+    - Create/Update event:
+      - For all the `ConfigMap`s referenced by the resource:
+        1. Update the dependency graph (If the resource is using a ConfigMap, mark the ConfigMap that the resource is using it)
+        2. If there are any updates to the ConfigMap in the dependency graph, add the key for the ConfigMap to add-lien-queue
+    - Delete event:
+      - For all the `ConfigMap`s referenced by the resource:
+        1. Update the dependency graph (If the resource is using a ConfigMap, unmark the ConfigMap that the resource is using it)
+        2. If there are no resource using the ConfigMap in the dependency graph, add the key for the ConfigMap to delete-lien-queue
+  - The controller periodically checks `Pod`, `Deployment`, `StatefulSet`, `DaemonSet`, and `CronJob`
+    - For each resource:
+      1. Update the dependency graph (If the resource is using a ConfigMap, mark the ConfigMap that the resource is using it)
+      2. If there are any updates to the ConfigMap in the dependency graph, add the key for the ConfigMap to add-lien-queue
+  - The controller periodically checks `ConfigMap`
+    - For each `ConfigMap`:
+      - Check if the `ConfigMap` have `kubernetes.io/enable-configmap-protection: "yes"` annotation:
+        - If yes:
+          1. Check API server if the resources marked as using the ConfigMap in the dependency graph are still using the ConfigMap, and update the graph
+          2. Check if there is no using resources in the graph any more
+            - If yes: Add the key for the ConfigMap to delete-lien-queue
+            - Otherwise: Add the key for the ConfigMap to add-lien-queue
+        - Otherwise: Add the key for the ConfigMap to delete-lien-queue
+  - The controller gets a key for a `ConfigMap` from add-lien-queue:
+    1. If the `ConfigMap` doesn't have `kubernetes.io/enable-configmap-protection: "yes"` annotation, do nothing
+    2. Check API server if the `ConfigMap` is actually used by one of the resources marked in the dependency graph, and if used, add `kubernetes.io/configmap-protection` lien to the `ConfigMap`
+  - The controller gets a key for a `ConfigMap` from delete-lien-queue:
+    1. If the `ConfigMap` doesn't have `kubernetes.io/enable-configmap-protection: "yes"` annotation, delete `kubernetes.io/configmap-protection` lien from the `ConfigMap`
+    2. Check API server if the `ConfigMap` is actually not used by any of the resources, and if not used, delete `kubernetes.io/configmap-protection` lien from the `ConfigMap`
+
+Note that
+  - Periodical checks for referencing resources are needed for handling a case that the annotation is added after referecing resources are created,
```

Comment:
If we would be watching, then periodic checks aren't really needed, as the event will be triggered by change of secret/configmap.